resource/file/util.go

```diff
@@ -0,0 +1,180 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package file utilites for managing file resources
+package file
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"strconv"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+// ModeType sets the higher-order bits of an os.FileMode based on the file type
+func ModeType(mode uint32, filetype Type) uint32 {
+	m := os.FileMode(mode)
+	switch filetype {
+	case TypeFile:
+		m &^= os.ModeType //clear bits if they are set
+	case TypeDirectory:
+		m |= os.ModeDir
+	case TypeSymlink:
+		m |= os.ModeSymlink
+	}
+	return uint32(m)
+}
+
+// GetType determines the file type and returns a Type
+func GetType(fi os.FileInfo) (Type, error) {
+	switch mode := fi.Mode(); {
+	case mode.IsRegular():
+		return TypeFile, nil
+	case mode.IsDir():
+		return TypeDirectory, nil
+	case mode&os.ModeSymlink == os.ModeSymlink:
+		return TypeSymlink, nil
+	default:
+		return TypeNone, fmt.Errorf("unsupported filetype for %s", fi.Name())
+	}
+}
+
+// UID returns the Unix Uid of a File
+func UID(fi os.FileInfo) (int, error) {
+	if stat, ok := fi.Sys().(*syscall.Stat_t); ok && stat != nil {
+		return int(stat.Uid), nil
+	}
+	return 0, fmt.Errorf("UID: stat.Sys failed")
+}
+
+// GID returns the Unix Gid of a File
+func GID(fi os.FileInfo) (int, error) {
+	if stat, ok := fi.Sys().(*syscall.Stat_t); ok && stat != nil {
+		return int(stat.Gid), nil
+	}
+	return 0, fmt.Errorf("GID: stat.Sys failed")
+}
+
+// UserInfo Returns the (unix user.User, error) of a file
+func UserInfo(fi os.FileInfo) (*user.User, error) {
+	uid, err := UID(fi)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get uid for file")
+	}
+
+	user, err := user.LookupId(strconv.Itoa(uid))
+	if err != nil {
+		return nil, fmt.Errorf("unable to get username for uid %d", uid)
+	}
+
+	return user, nil
+
+}
+
+// GroupInfo returns the (Unix user.Group of a File, error)
+func GroupInfo(fi os.FileInfo) (*user.Group, error) {
+	gid, err := GID(fi)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get uid for file")
+	}
+
+	group, err := user.LookupGroupId(strconv.Itoa(gid))
+	if err != nil {
+		return &user.Group{}, fmt.Errorf("unable to get group name for gid %d", gid)
+	}
+
+	return group, nil
+
+}
+
+//given two users, decide which one to use
+//returns true if a change is required
+func desiredUser(f, actual *user.User) (userInfo *user.User, changed bool, err error) {
+	if f == nil || f.Username == "" {
+		if actual == nil || actual.Username == "" { // if neither is set, use the effective uid of the process
+			userInfo, err = user.LookupId(strconv.Itoa(os.Geteuid()))
+			if err != nil {
+				return &user.User{}, false, errors.Wrapf(err, "unable to set default username %s", f.Name)
+			}
+			return userInfo, true, err
```

Comment:
Probably better to return nil instead of err to be more clear...it would be good to change throughout the code.